apps/test/unit/templates/sectionProgress/ProgressTableLevelBubbleTest.js

```diff
@@ -0,0 +1,175 @@
+import {assert, expect} from '../../../util/reconfiguredChai';
+import React from 'react';
+import {shallow} from 'enzyme';
+import ProgressTableLevelBubble from '@cdo/apps/templates/sectionProgress/progressTables/ProgressTableLevelBubble';
+import color from '@cdo/apps/util/color';
+import FontAwesome from '@cdo/apps/templates/FontAwesome';
+import {LevelStatus, LevelKind} from '@cdo/apps/util/sharedConstants';
+import i18n from '@cdo/locale';
+
+const TITLE = '1';
+
+const defaultProps = {
+  levelStatus: LevelStatus.not_tried,
+  levelKind: LevelKind.level,
+  disabled: false,
+  title: TITLE,
+  url: '/foo/bar'
+};
+
+const borderColors = {
+  [LevelStatus.locked]: color.lighter_gray,
+  [LevelStatus.not_tried]: color.lighter_gray,
+  [LevelStatus.attempted]: color.level_perfect,
+  [LevelStatus.passed]: color.level_perfect,
+  [LevelStatus.perfect]: color.level_perfect
+};
+const backgroundColors = {
+  [LevelStatus.locked]: color.level_not_tried,
+  [LevelStatus.not_tried]: color.level_not_tried,
+  [LevelStatus.attempted]: color.level_not_tried,
+  [LevelStatus.passed]: color.level_passed,
+  [LevelStatus.perfect]: color.level_perfect
+};
+const assessmentBorders = {
+  [LevelStatus.locked]: color.lighter_gray,
+  [LevelStatus.not_tried]: color.lighter_gray,
+  [LevelStatus.attempted]: color.level_submitted,
+  [LevelStatus.submitted]: color.level_submitted,
+  [LevelStatus.completed_assessment]: color.level_submitted
+};
+const assessmentBackgrounds = {
+  [LevelStatus.locked]: color.level_not_tried,
+  [LevelStatus.not_tried]: color.level_not_tried,
+  [LevelStatus.attempted]: color.level_not_tried,
+  [LevelStatus.submitted]: color.level_submitted,
+  [LevelStatus.completed_assessment]: color.level_submitted
+};
+
+/**
+ * Helper function to retrieve the style object of a rendered bubble with the
+ * provided status. We use a "bonus" bubble to make it easy to find the
+ * FontAwesome content node. The content is in a positioning container, and the
+ * parent of that container is the actual "bubble" node with stylized border
+ * and background, so we want the style of the content's grandparent.
+ */
+function bubbleContainerStyleForStatus(status, propOverrides = {}) {
+  const wrapper = shallow(
+    <ProgressTableLevelBubble
+      {...defaultProps}
+      {...propOverrides}
+      levelStatus={status}
+      bonus={true}
+    />
+  );
+  return wrapper
+    .find(FontAwesome)
+    .at(0)
+    .parents()
+    .at(1)
+    .props().style;
+}
+
+describe('ProgressTableLevelBubble', () => {
```

Comment:
Not sure what you ended up thinking about these unit tests, but these seem like they're useful.

Reply:
i tried to make them feel less fragile than some of the ProgressBubble tests by using my `bubbleContainerStyleForStatus` helper to find the relevant styled note relative to the content, rather than relying on knowing which level down from the top node it should be found at, but imo that's a pretty minor improvement.

my general feeling on these is still that they're not very useful because i think they are mostly verifying that the code they're testing hasn't _changed_, not that it hasn't _broke_. it seems to me like our tests should mostly be helping us catch regressions rather than simply making sure that when we modify code we remember to update all the tests that touch it, and i'm having a hard time seeing how tests like these are very helpful in that regard.

also, they're not actually validating what they claim to be validating, which is that the rendered component looks how it should. checking properties of DOM nodes doesn't actually tell us much about how something looks -- it would be easy for those properties to be overridden somewhere else in the tree, or maybe the node we're checking isn't actually visible for some other reason, etc..

i would love it if we could figure out a way to test components that is more like our eyes tests (i.e. verifies that what's rendered matches some baseline), because then we could verify that code changes didn't cause regressions even when they change the structure of the DOM tree in a way that these unit tests would find problematic.

Reply:
another point worth mentioning here is how much testing harness we want to include in our actual code. `ProgressBubble` uses explicit CSS class names to make it easier to validate the presence/absence of specific nodes in the rendered tree, which makes it easier to test logic (e.g. "if i pass X property, is Y node rendered"). i personally find that messy so i didn't use it here, but maybe that is the best way to make this kind of testing more useful.

imo a better way to accomplish the same thing would be to break our components down into smaller pieces, so anyplace we would previously have added a CSS class name to find a node, we would instead break that node out into its own component, which enzyme can easily find.

in this case, that would mean breaking out separate components for "unplugged", "small circle", "large circle", and "large diamond".

we can discuss this when i schedule a call to discuss testing strategy for the remaining refactor components, and then again in more depth in H1 with the broader team.